File: src/visual-graph/builder/ensure-beyond-depth-stub.ts
import { NODE_KIND } from "../node-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualNode } from "../visual-node.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import type { BuildState } from "./build-state.js";
import { sanitize } from "./sanitize.js";

// Return (creating on first call) the single BeyondDepth stub for a
// given visible ancestor subgraph. Multiple anonymous collapsed children
// of the same surviving outer container funnel into the same placeholder,
// so the rendered graph shows one `((...))` boundary marker per visible
// parent instead of one per hidden child.
export function ensureBeyondDepthStub(
  parentSubgraph: VisualSubgraph,
  state: BuildState,
): string {
  const existing = state.beyondDepthStubByParent?.get(parentSubgraph.id);
  if (existing !== undefined) {
    return existing;
  }
  const stubId = `beyond_depth_${sanitize(parentSubgraph.id)}`;
  const node = {
    type: VISUAL_ELEMENT_TYPE.Node,
    id: stubId,
    kind: NODE_KIND.LegacyBeyondDepth,
    name: "...",
    line: parentSubgraph.line,
    endLine: parentSubgraph.endLine,
    isJsxElement: false,
    unused: false,
  } satisfies VisualNode;
  parentSubgraph.elements.push(node);
  state.beyondDepthStubByParent?.set(parentSubgraph.id, stubId);
  return stubId;
}

File: crates/unsnarl-visual-graph/src/builder/ensure_beyond_depth_stub.rs
//! Return (creating on first call) the single BeyondDepth stub
//! associated with a visible ancestor subgraph. Multiple anonymous
//! collapsed children of the same surviving outer container funnel
//! into the same `((...))` placeholder so the rendered graph shows
//! one boundary marker per visible parent instead of one per hidden
//! child.

use crate::visual_node::SyntheticVisualNode;

use super::arena::{BuildArena, Container, ElementHandle, SubgraphIdx};
use super::sanitize::sanitize;
use super::state::BuildState;

pub fn ensure_beyond_depth_stub(
    arena: &mut BuildArena,
    state: &mut BuildState,
    parent: SubgraphIdx,
) -> String {
    let parent_id = arena.subgraph(parent).descriptor.id().to_string();
    if let Some(existing) = state.beyond_depth_stub_by_parent.get(&parent_id) {
        return existing.clone();
    }
    let stub_id = format!("beyond_depth_{}", sanitize(&parent_id));
    let line = arena.subgraph(parent).descriptor.line();
    let end_line = arena.subgraph(parent).descriptor.end_line();
    let mut n = SyntheticVisualNode::beyond_depth(stub_id.clone(), line);
    n.end_line = end_line;
    let node = n.into();
    let node_idx = arena.push_node(node);
    arena.append_child(Container::Subgraph(parent), ElementHandle::Node(node_idx));
    state
        .beyond_depth_stub_by_parent
        .insert(parent_id, stub_id.clone());
    stub_id
}
